File: game/Assets/Scripts/Network/Messages/MessageToServer.cs
using System;
using System.Collections.Generic;

namespace Network
{

    [Serializable]
    public class MessageToServer
    {
        public enum MessageType
        {
            // Send list of player actions to the server.
            ACTIONS = 0,
            // Request the server send a state synch.
            STATE_SYNC_REQUEST,

            // Used to join a new game/other game management actions.
            ROOM_MANAGEMENT,
            // Used by the leader to send objectives (tasks) to the follower.
            OBJECTIVE,
            // Used by the follower to signal that an objective has been completed.
            OBJECTIVE_COMPLETE,
            TURN_COMPLETE,
            TUTORIAL_REQUEST,
            PONG,
            LIVE_FEEDBACK,
            CANCEL_PENDING_OBJECTIVES,
            GOOGLE_AUTH,
            USER_INFO,
        }

        // These fields are populated for every packet.
        public string transmit_time;  // Transmission time of this message in ISO 8601 format.

        // Depending on the type, One of the following may be populated.
        public MessageType type;

        public List<Action> actions;
        public RoomManagementRequest room_request;
        public ObjectiveMessage objective;
        public ObjectiveCompleteMessage objective_complete;
        public TurnComplete turn_complete;
        public TutorialRequest tutorial_request;
        public Pong pong;
        public LiveFeedback live_feedback;
        public GoogleAuth google_auth;
    }

}  // namespace Network

File: server/messages/message_to_server.py
""" Defines message structure sent to server.  """

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional

from mashumaro import pass_through
from mashumaro.mixins.json import DataClassJSONMixin

from server.messages.action import Action
from server.messages.google_auth import GoogleAuth
from server.messages.live_feedback import LiveFeedback
from server.messages.objective import ObjectiveCompleteMessage, ObjectiveMessage
from server.messages.pong import Pong
from server.messages.replay_messages import ReplayRequest
from server.messages.rooms import RoomManagementRequest
from server.messages.turn_state import TurnComplete
from server.messages.tutorials import TutorialRequest


class MessageType(Enum):
    ACTIONS = 0
    STATE_SYNC_REQUEST = 1
    ROOM_MANAGEMENT = 2
    OBJECTIVE = 3
    OBJECTIVE_COMPLETED = 4
    TURN_COMPLETE = 5
    TUTORIAL_REQUEST = 6
    PONG = 7
    LIVE_FEEDBACK = 8
    CANCEL_PENDING_OBJECTIVES = 9
    GOOGLE_AUTH = 10
    USER_INFO = 11
    REPLAY_REQUEST = 12
    SCENARIO_REQUEST = 13


@dataclass(frozen=True)
class MessageToServer(DataClassJSONMixin):
    transmit_time: datetime = field(
        metadata={"deserialize": "pendulum", "serialize": pass_through}
    )
    type: MessageType = MessageType.ACTIONS
    actions: Optional[List[Action]] = None
    room_request: Optional[RoomManagementRequest] = None
    objective: Optional[ObjectiveMessage] = None
    objective_complete: Optional[ObjectiveCompleteMessage] = None
    turn_complete: Optional[TurnComplete] = None
    tutorial_request: Optional[TutorialRequest] = None
    pong: Optional[Pong] = None
    live_feedback: Optional[LiveFeedback] = None
    google_auth: Optional[GoogleAuth] = None
    replay_request: Optional[ReplayRequest] = None
